node/service/src/lib.rs

```diff
@@ -266,34 +270,67 @@ fn new_partial<RuntimeApi, Executor>(config: &mut Configuration) -> Result<
 	})
 }
 
-fn real_overseer<S: SpawnNamed>(
+fn make_metrics<M: Metrics + Default>(registry: Option<&Registry>) -> Result<M, ServiceError> {
+	Ok(match registry {
+		None => M::default(),
+		Some(registry) => M::try_register(registry)?,
+	})
+}
+
+fn real_overseer<Spawner, RuntimeClient>(
 	leaves: impl IntoIterator<Item = BlockInfo>,
-	prometheus_registry: Option<&Registry>,
-	s: S,
-) -> Result<(Overseer<S>, OverseerHandler), ServiceError> {
+	keystore: KeyStorePtr,
+	runtime_client: RuntimeClient,
+	availability_config: AvailabilityConfig,
+	network_service: Arc<sc_network::NetworkService<Block, Hash>>,
+	authority_discovery: AuthorityDiscoveryService,
+	registry: Option<&Registry>,
+	spawner: Spawner,
+) -> Result<(Overseer<Spawner>, OverseerHandler), ServiceError>
+where
+	RuntimeClient: ProvideRuntimeApi<Block> + HeaderBackend<Block>,
+	RuntimeClient::Api: ParachainHost<Block>,
+	Spawner: SpawnNamed + Clone + Unpin,
+{
+	use polkadot_availability_distribution::AvailabilityDistributionSubsystem;
+	use polkadot_node_core_av_store::AvailabilityStoreSubsystem;
+	use polkadot_availability_bitfield_distribution::BitfieldDistribution as BitfieldDistributionSubsystem;
+	use polkadot_node_core_bitfield_signing::BitfieldSigningSubsystem;
+	use polkadot_node_core_backing::CandidateBackingSubsystem;
+	use polkadot_node_core_candidate_selection::CandidateSelectionSubsystem;
+	use polkadot_node_core_candidate_validation::CandidateValidationSubsystem;
+	use polkadot_node_core_chain_api::ChainApiSubsystem;
+	use polkadot_node_collation_generation::CollationGenerationSubsystem;
+	use polkadot_collator_protocol::CollatorProtocolSubsystem;
+	use polkadot_network_bridge::NetworkBridge as NetworkBridgeSubsystem;
+	use polkadot_pov_distribution::PoVDistribution as PoVDistributionSubsystem;
+	use polkadot_node_core_provisioner::ProvisioningSubsystem as ProvisionerSubsystem;
+	use polkadot_node_core_runtime_api::RuntimeApiSubsystem;
+	use polkadot_statement_distribution::StatementDistribution as StatementDistributionSubsystem;
+
 	let all_subsystems = AllSubsystems {
-		candidate_validation: DummySubsystem,
-		candidate_backing: DummySubsystem,
-		candidate_selection: DummySubsystem,
-		statement_distribution: DummySubsystem,
-		availability_distribution: DummySubsystem,
-		bitfield_signing: DummySubsystem,
-		bitfield_distribution: DummySubsystem,
-		provisioner: DummySubsystem,
-		pov_distribution: DummySubsystem,
-		runtime_api: DummySubsystem,
-		availability_store: DummySubsystem,
-		network_bridge: DummySubsystem,
-		chain_api: DummySubsystem,
-		collation_generation: DummySubsystem,
-		collator_protocol: DummySubsystem,
+		availability_distribution: AvailabilityDistributionSubsystem::new(keystore.clone(), make_metrics(registry)?),
```

Comment:
Maybe we should change the signature of all `Subsystem::new` to accept `Option<&PrometheusRegistry>` instead of a metric? This way the metric type could be made private, also it would be in line with `CollatorProtocolSubsystem::new`.

Reply:
After chasing this down for a while, I've had to give it up: `struct JobManager` has to have a field `Job::Metrics` to work properly; replacing that with the registry information both forces it to be parametrized by the registry's lifetime, and forces re-registration each time the metrics are updated. We could potentially make some of the `Metrics` types private which don't use the subsystem-util, but in this case, I'd prefer that the interface is consistent more than that it exposes the fewest possible types.